src/interval.jl

```diff
@@ -364,6 +385,206 @@ true
+    if isempty(x)
+        return false
+    elseif isempty(y)
+        return true
+    else
+        return isless(by(first(x)), by(first(y)))
+    end
+end
+
+function first_is_equal(x, y, by=identity)
+    if isempty(x)
+        return false
+    elseif isempty(y)
+        return false
+    else
+        return isequal(by(first(x)), by(first(y)))
+    end
+end
+
+function first_is_closed(x, by=identity)
+    if isempty(x)
+        true
+    else
+        return isclosed(by(first(x)))
+    end
+end
+
+first_is_left(x; by=identity) = isempty(x) ? false : isleft(by(first(x)))
+endpoint_dir_type(::Type) = NaN # a value unequal to itself
+endpoint_dir_type(::Type{<:HalfOpenEndpoint{T, B}}) where {T, B} = B
+endpoint_dir_type(::Type{<:AbstractArray{T}}) where T = endpoint_dir_type(T)
+endpoint_dir_type(x::T) where T = endpoint_dir_type(T)
+
+#     mergesets(op, x, y)
+#
+# `mergesets` is the primary internal function implementing set operations (see
+# below for its usage). It iterates through the left and right endpoints in x
+# and y, in order from lowest to highest. The implementation is based on the
+# insight that we can make a decision to include or exclude all points after a
+# given endpoint (based on `op`) and that decision will remain unchanged moving
+# left to right along the real-number line until we encounter a new endpoint.
+#
+# For each endpoint, we determine two things: 
+#   1. whether subsequent points should be included in the merge operation or
+#        not (based on its membership in both `x` and `y`) by using `op`
+#   2. whether the next step will define a left (start including) or right
+#        endpoint (stop includeing)
+#
+# Then, we decide to add a new endpoint if 1 and 2 match (i.e. "should include"
+# points will create a time point when the next point will start including
+# points).
+#
+# A final issue is handling the closed/open nature of each endpoint. In the
+# general case, we have to track whether to keep the endpoint (closed) or not
+# (open) separately. Keeping the endpoint may require we keep a singleton
+# endpoint ([1,1]) such as when to closed endpoints intersect with one another
+# (e.g. (0, 1] ∩ [1, 2))
+#
+function mergesets(op, x, y)
+    result = promote_type(eltype(x), eltype(y))[]
+    sizehint!(result, length(x) + length(y))
+
+    # to start, points are not included (until we see the starting endpoint of a set)
+    inresult = false
+    inx = false
+    iny = false
+
+    # we do not have to track the endpoints if we're using HalfOpenEndpoint
+    # objects 
+    # NOTE: since this statement relies only on types, the branches using this
+    # flag should compile away
+    track_endpoint = endpoint_dir_type(x) != endpoint_dir_type(y)
+
+    while !(isempty(x) && isempty(y))
+        # println("-------------------")
+        #=@show=# t = first_is_less(x, y) ? first(x) : first(y)
+        #=@show=# x_isless = first_is_less(x, y)
+        #=@show=# x_equal = first_is_equal(x, y)
```

Comment:
Why not just call `iterate` or `Iterators.peel` and then do these checks?

Reply:
Sometimes one of the iterables is empty; I find this form more readible than introducing the branches to check for `nothing`.

Reply:
I guess the other point to make here is that we only want to peel the values when they are "first" (i.e. less than or equal to the value in the other list), so you don't want to peel before that.